src/validation.cpp

```diff
@@ -1013,18 +1038,33 @@ bool MemPoolAccept::Finalize(const ATMPArgs& args, Workspace& ws)
                 (int)entry->GetTxSize() - (int)ws.m_conflicting_size);
         ws.m_replaced_transactions.push_back(it->GetSharedTx());
     }
-    m_pool.RemoveStaged(ws.m_all_conflicting, false, MemPoolRemovalReason::REPLACED);
+    if (ws.m_witness_replace) {
+        // Only remove the direct conflict; preserve and update the descendants.
+        m_pool.RemoveStaged(ws.m_iters_conflicting, /* updateDescendants */ true, MemPoolRemovalReason::REPLACED);
+    } else {
+        m_pool.RemoveStaged(ws.m_all_conflicting, /* updateDescendants */ false, MemPoolRemovalReason::REPLACED);
+    }
 
     // This transaction should only count for fee estimation if:
     // - it's not being re-added during a reorg which bypasses typical mempool fee limits
     // - the node is not behind
     // - the transaction is not dependent on any other transactions in the mempool
     // - it's not part of a package. Since package relay is not currently supported, this
-    // transaction has not necessarily been accepted to miners' mempools.
-    bool validForFeeEstimation = !bypass_limits && !args.m_package_submission && IsCurrentForFeeEstimation(m_active_chainstate) && m_pool.HasNoInputsOf(tx);
+    //   transaction has not necessarily been accepted to miners' mempools
+    // - this is not a witness replacement. Miners may not have witness replacement implemented
+    //   (yet). TODO: re-include witness replacements in v25.0
```

Comment:
While I've seen protocol proposals leveraging same-txid-different-witness, I don't know any of them which is effectively deployed now (?). If they're and their usages is high enough to lead to a significant rate of witness conflicts, by that time I would expect this new replacement policy to be widely deployed in the miner ecosystem.

Of course, the majority of miners might run a different policy than the default Core's one. Though, until the contrary is corroborated by periodic probing of miner's mempools, I think we can reasonably hold the assumption they're.

So to simplify, I lean to account witness replacement in fee estimation as early as this PR.

What do you think ?

Reply:
The reasoning was based on #9519 - adding a witness replacement to the fee estimator isn't really accurate unless miners are also allowing witness replacements. I think it's reasonable to wait 1 or 2 releases for miners to update their policies? We just need to not forget to remove this bool. I imagine it'll take some time for it to be used, anyway.